ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -548,20 +553,29 @@ private IndexValue findKey(StoreKey key, FileSpan fileSpan, IndexEntryType type,
           Iterator<IndexValue> it = values.descendingIterator();
           while (it.hasNext()) {
             IndexValue value = it.next();
+            foundValues.add(value);
             logger.trace("Index : {} found value offset {} size {} ttl {}", dataDir, value.getOffset(), value.getSize(),
                 value.getExpiresAtMs());
-            if (type.equals(IndexEntryType.ANY)) {
-              retValue = value;
-              break;
-            } else if (type.equals(IndexEntryType.DELETE) && value.isFlagSet(IndexValue.Flags.Delete_Index)) {
+            if (types.contains(IndexEntryType.DELETE) && value.isFlagSet(IndexValue.Flags.Delete_Index)) {
               retValue = value;
               break;
-            } else if (type.equals(IndexEntryType.PUT) && !value.isFlagSet(IndexValue.Flags.Delete_Index)) {
+            } else if (types.contains(IndexEntryType.PUT) && !value.isFlagSet(IndexValue.Flags.Delete_Index)
+                && !value.isFlagSet(IndexValue.Flags.Ttl_Update_Index)) {
               retValue = value;
               break;
             }
+            // note that it is not possible for a TTL update record to exist for a key but not have a PUT or DELETE
+            // record.
           }
           if (retValue != null) {
+            // merge entries if required to account for updated fields
+            IndexValue latest = foundValues.get(0);
+            if (latest.isFlagSet(IndexValue.Flags.Ttl_Update_Index) && !retValue.isFlagSet(
+                IndexValue.Flags.Ttl_Update_Index)) {
```

Comment:
Isn't the following check sufficient and more readable?

```java
if (latest != retValue && latest.isFlagSet(Ttl_Update_Index)) {
   ...
}
```

Reply:
Actually I find this more readable in the sense "the latest has the ttl update flag set and the return candidate doesn't, so let's make a change"